MapboxNavigation/InstructionsCardCollection.swift

```diff
@@ -0,0 +1,332 @@
+        
+        addSubviews()
+        setConstraints()
+        
+        view.clipsToBounds = false
+        topPaddingView.backgroundColor = .clear
+    }
+    
+    func addSubviews() {
+        [topPaddingView, instructionCollectionView].forEach(view.addSubview(_:))
+    }
+    
+    func setConstraints() {
+        let topPaddingConstraints: [NSLayoutConstraint] = [
+            topPaddingView.topAnchor.constraint(equalTo: view.topAnchor),
+            topPaddingView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+            topPaddingView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+            topPaddingView.bottomAnchor.constraint(equalTo: view.safeTopAnchor),
+            ]
+        
+        NSLayoutConstraint.activate(topPaddingConstraints)
+        
+        let instructionCollectionViewContraints: [NSLayoutConstraint] = [
+            instructionCollectionView.topAnchor.constraint(equalTo: topPaddingView.bottomAnchor),
+            instructionCollectionView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+            instructionCollectionView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+            instructionCollectionView.heightAnchor.constraint(equalToConstant: cardSize.height),
+            instructionCollectionView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
+        ]
+        
+        NSLayoutConstraint.activate(instructionCollectionViewContraints)
+    }
+    
+    func reloadDataSource() {
+        if isInPreview {
+            if previewIndexPath.row < steps!.endIndex - 1 {
+                updateVisibleInstructionCards(at: [previewIndexPath], previewEnabled: true)
+            } else {
+                instructionCollectionView.reloadData()
+            }
+        } else {
+            if currentStepIndex == nil, let progress = routeProgress {
+                currentStepIndex = progress.currentLegProgress.stepIndex
+                instructionCollectionView.reloadData()
+            } else if let progress = routeProgress, let stepIndex = currentStepIndex, stepIndex != progress.currentLegProgress.stepIndex {
+                currentStepIndex = progress.currentLegProgress.stepIndex
+                instructionCollectionView.reloadData()
+            } else {
+                updateVisibleInstructionCards(at: instructionCollectionView.indexPathsForVisibleItems)
+            }
+        }
+    }
+    
+    func updateVisibleInstructionCards(at indexPaths: [IndexPath], previewEnabled: Bool = false) {
+        guard let distances = distancesFromCurrentLocationToManeuver else { return }
+        for index in indexPaths.startIndex..<indexPaths.endIndex {
+            let indexPath = indexPaths[index]
+            if let container = instructionContainerView(at: indexPath), indexPath.row < distances.endIndex {
+                let distance = distances[indexPath.row]
+                container.updateInstructionCard(distance: distance, previewEnabled: previewEnabled)
```

Comment:
In https://github.com/mapbox/mapbox-navigation-ios/pull/2149#discussion_r295589920, I pointed out that it’s quite inefficient to calculate the route-line distance from the current step to every remaining maneuver on every location update (once a second). In turns out this was by design: each previewed card (and indeed the follow-on card peeking in from the side) displays a distance measured from the current location at all times, resulting in monotonically increasing distances.

I hope we’ll be able to reconsider this design, which is unfortunate not only for its inherently poor performance but also unintuitiveness to the end user. But for now, there isn’t a good way around performing all those calculations.